keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+  expectedEvacuationFinishTime: "2024-04-22T21:41:32Z" # now + terminationGracePeriodSeconds:
+  failedEvictionCounter: 0
+  message: "critical-ds is terminating the pod due to node maintenance (OS upgrade)."
+  conditions: []
+```
+
+Once the pod is terminated and removed from the node, it should not be re-scheduled on the node by
+the deamonset controller until the node maintenance is complete.
+
+### Kubelet: Graceful Node Shutdown
+
+The current Graceful Node Shutdown feature has a couple of drawbacks when compared to
+NodeMaintenance:
+- It is application agnostic as it only provides a static grace period before the shutdown based on
+  priority. This does not always give the application enough time to react and can lead to data
+  loss or application availability loss.
+- The DaemonSet pods may be running important services (critical priority) that should be available
+  even during part of the shutdown. The daemon set controller does not have the observability of
+  the kubelet shutdown procedure and cannot infer which DaemonSets should stop running. The
+  controller needs to know which DaemonSets should run on each node with which priorities and
+  reconcile accordingly.
+
+To support these use cases we could introduce a new configuration option to the kubelet called
+`preferNodeMaintenanceDuringGracefulShutdown`.
+
+This would result in the following behavior:
+
+When a shutdown is detected, the kubelet would create a NodeMaintenance object for that node.
+Then it would block the shutdown indefinitely, until all the pods are terminated. The kubelet
+could pass the priorities from the `shutdownGracePeriodByPodPriority` to the NodeMaintenance,
+just without the `shutdownGracePeriodSeconds`. This would give applications a chance to react and
+gracefully leave the node without a timeout. [Pod Selection](#pod-selection) would ensure that user
+workloads are terminated first and critical pods are terminated last. 
+
+By default, all user workloads will be asked to terminate at once. The Evacuation API ensures that
+an evacuator is selected or an eviction API is called. This should result in a fast start of a pod
+termination.NodeMaintenance could then be used even by spot instances.
+
+The NodeMaintenance object could survive kubelet restarts, and the kubelet would always know if the
+node is under shutdown (maintenance). The cluster admin would have to remove the NodeMaintenance
+object after the node restart to indicate that the node is healthy and can run pods again. Admins
+are expected to deal with the lifecycle of planned NodeMaintenances, so reacting to the unplanned
+one should not be a big issue.
+
+If there is no connection to the apiserver (apiserver down, network issues, etc.) and the
+NodeMaintenance object cannot be created, we would fall back to the original behavior of Graceful
+Node Shutdown feature. If the connection is restored, we would stop the Graceful Node Shutdown and
```

Comment:
It might be too late to fallback to something else. We should attempt to add NodeMaintenance object, should it be missing, but it might be too late to stop Graceful Node Shutdown, which might be too far gone, so to speak. This would be a one-way? hard? fallback.

Reply:
Alternatively, perhaps then doing nothing or refusing to carry-out a shutdown would be a safer option?

That said, depends on how the shutdown was triggered: if it was due to external factors (which would be the default modus operandi for which Graceful Node Shutdown has been designed), then it would be potentially unsafe to interrupt the ongoing shutdown, or there might be not enough time, etc.

Reply:
It would be too late for the pods that are already terminating. But it might be useful for the running non-terminated pods, right?

It seems to me that by default it would be good to carry out the Graceful Node Shutdown in the disconnected scenario. At least to have a softer landing in bad situations. I wonder if it would make sense to make this behavior configurable?

Reply:
AIUI, DNM is about marking the node as needing maintenance and barely overlaps with node shutdown handling.

The only overlap is that both things trigger a node drain; but other things could also trigger that, such as node problem detector deciding the node is irreparably faulty.